### PR TITLE
Add IRCv3 account-notify and extended-join

### PR DESCRIPTION
**Pull request by vanosg**
Test cases demonstrating functionality (if applicable):
```
.tcl bind acnt * * foo
Tcl: *
.tcl proc foo {nick user hand acnt} {putlog "nick $nick user $user hand $hand acnt $acnt"}
Tcl: 
[05:52:39] [@] :user1!user1@oragono.io ACCOUNT user
[05:52:39] user1@oragono.io has logged into account user
[05:52:39] nick user1 user user1@oragono.io hand admin acnt user
```

### CODE REVIEW

**Comment by vanosg** [comment]
Updated demo:
```
.tcl bind acnt * * foo
Tcl: * 
.tcl proc foo {nick user hand chan account} {putlog "nick $nick user $user hand $hand chan $chan account $account"}
Tcl: 
.msg nickserv identify myaccount password
[04:27:51] BeerBot!user@foo.io has logged into account myaccount
[04:27:51] nick BeerBot user user@foo.io hand * chan #eggdroptest account myaccount
[04:27:51] -NickServ (NickServ@localhost)- You're now logged in as myaccount

.tcl getaccount beerbot #eggdroptest
Tcl: myaccount

## And an unauthenticated user:
.tcl getaccount otheruser  #eggdroptest
Tcl: 
```

**Comment by vanosg** [comment]
Demo of extended-join:
```
.tcl cap enabled
Tcl: extended-join

.tcl proc foo3 {nick user hand chan account} {putlog "3: nick is $nick user is $user hand is $hand chan is $chan account is $account"}
Tcl: 
.tcl proc foo2 {nick user hand chan account} {putlog "2: nick is $nick user is $user hand is $hand chan is $chan account is $account"}
Tcl: 
.tcl proc foo1 {nick user hand chan account} {putlog "1: nick is $nick user is $user hand is $hand chan is $chan account is $account"}
Tcl: 

.tcl bind acnt * "#eggdroptest *" foo1
Tcl: #eggdroptest *
.tcl bind acnt * "* myuser!*@*" foo2
Tcl: * myuser!*@*
.tcl bind acnt * "* myaccount" foo3
Tcl: * myaccount

### <user already logged in as 'myaccount' joins channel>
[05:36:32] 3: nick is myuser user is foo@server.io hand is * chan is #eggdroptest account is myaccount
[05:36:32] 2: nick is myuser user is foo@server.io hand is * chan is #eggdroptest account is myaccount
[05:36:32] 1: nick is myuser user is foo@server.io hand is * chan is #eggdroptest account is myaccount

.tcl getaccount myuser
Tcl: myaccount
```

**Comment by vanosg** [comment]
Test for isloggedin:
```
.tcl isloggedin geo
Tcl: 0
### <user logs in, with account-notify active>
[03:34:01] geo!foo@bar.io has logged into account geo
.tcl isloggedin geo
Tcl: 1
```

**Comment by vanosg** [comment]
Added account2nick and account2handle Tcl commands:

```
 NICKNAME  HANDLE    ACCOUNT    JOIN   IDLE  USER@HOST
 BeerBot   *                    ---         <- it's me!
 Geo-      Moo       Geo        ---   N      foo@unaffiliated/geo
End of channel info.


.tcl account2hand geo
Tcl: Moo
.tcl account2nick geo
Tcl: Geo-
```

File: src/mod/server.mod/tclserv.c [hunk 1]
```diff
@@ -162,6 +162,65 @@ static int tcl_puthelp STDVAR
...
+    chan = chanset;
+  }
+  while (chan && (thechan == NULL || thechan == chan)) {
+    if ((m = ismember(chan, argv[1]))) {
+      Tcl_AppendResult(irp, m->account, NULL);
+      return TCL_OK;
+    }
+    chan = chan->next;
+  }
+  Tcl_AppendResult(irp, "0", NULL);
```
Comment (vanosg) [inline src/mod/server.mod/tclserv.c]: Change to ""

File: src/mod/server.mod/tclserv.c [hunk 2]
```diff
@@ -451,5 +510,7 @@ static tcl_cmds my_tcl_cmds[] = {
   {"tagmsg",     tcl_tagmsg},
   {"addserver",  tcl_addserver},
   {"delserver",  tcl_delserver},
+  {"getaccount", tcl_getaccount},
+  {"isloggedin", tcl_isloggedin},
```
Comment (vanosg) [inline src/mod/server.mod/tclserv.c]: Change to isidentified

File: src/mod/irc.mod/chan.c [hunk 3]
```diff
@@ -1811,6 +1823,16 @@ static int gotjoin(char *from, char *chname)
         m->user = u;
         m->flags |= STOPWHO;
 
+        if (extended_join) {
```
Comment (thommey) [inline src/mod/irc.mod/chan.c:1859]: this is dead code, right? now that check_tcl_account is removed this does nothing

File: src/mod/server.mod/servmsg.c [hunk 4]
```diff
@@ -1487,7 +1503,35 @@ static int handle_sasl_timeout()
...
+  struct userrec *u;
+  memberlist *m;
+  char *nick, *chname, mask[CHANNELLEN+UHOSTLEN+NICKMAX+2];
+
+  u = get_user_by_host(from);
+  nick = splitnick(&from);
+  for (chan = chanset; chan; chan = chan->next) {
+    chname = chan->dname;
+    if ((m = ismember(chan, nick))) {
+      strncpy (m->account, msg, sizeof m->account);
```
Comment (thommey) [inline src/mod/server.mod/servmsg.c]: strlcpy

File: src/mod/irc.mod/chan.c (hunk 3 above)
Comment (vanosg) [reply src/mod/irc.mod/chan.c:1859]: some of it... ;)

File: src/mod/irc.mod/chan.c [hunk 5]
```diff
@@ -1204,7 +1218,7 @@ static int got367(char *from, char *origmsg)
   char *ban, *who, *chname, buf[511], *msg;
   struct chanset_t *chan;
 
-  strncpy(buf, origmsg, 510);
+  strlcpy(buf, origmsg, 510);
   buf[510] = 0;
```
Comment (michaelortmann) [inline src/mod/irc.mod/chan.c:1222]: since we use strlcpy now, we can remove this null terminating no-op

**Pull request merged**